File: app-java/src/main/java/com/tritiumstudios/phasmophobiaevidencetool/android/data/viewmodels/ObjectivesViewModel.java
package com.tritiumstudios.phasmophobiaevidencetool.android.data.viewmodels;

import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModel;

import com.tritiumstudios.phasmophobiaevidencetool.android.activities.investigation.missions.data.MissionsData;

/**
 * ObjectivesViewModel class
 *
 * @author TritiumGamingStudios
 */
public class ObjectivesViewModel extends ViewModel {

    @Nullable
    private MissionsData.Objective[] objectivesSpinnerObjectives = null;
    @Nullable
    private boolean[] isObjectiveCompleted = null;

    @Nullable
    private String ghostName = null;

    private boolean responseState = false; // alone , group

    /*
     * Chosen spinner objectives
     */
    public void setObjectivesSpinnerObjectives(MissionsData.Objective[] objectiveSpinnerObjectives) {
        this.objectivesSpinnerObjectives = objectiveSpinnerObjectives;
    }

    public MissionsData.Objective[] getObjectivesSpinnerObjectives() {
        return objectivesSpinnerObjectives;
    }

    /*
     * Ghost name
     */
    public void setGhostName(String ghostName) {
        this.ghostName = ghostName;
    }

    public String getGhostName() {
        return ghostName;
    }

    /*
     * Response
     */
    public void setResponseState(boolean responseState) {
        this.responseState = responseState;
    }

    public boolean getResponseState() {
        return responseState;
    }

    /*
     * Objective Completed Buttons
     */
    public void setObjectiveCompletion(boolean[] isObjectiveCompleted) {
        this.isObjectiveCompleted = isObjectiveCompleted;
    }

    public boolean[] getObjectiveCompletion() {
        return isObjectiveCompleted;
    }

    /*
     * Reset
     */
    public void reset() {
        objectivesSpinnerObjectives = null;
        isObjectiveCompleted = null;
        ghostName = null;
        responseState = false;
    }

}

File: app-view/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/presentation/viewmodel/ObjectivesViewModel.kt
package com.tritiumgaming.phasmophobiaevidencepicker.presentation.viewmodel

import androidx.compose.runtime.collectAsState
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory.Companion.APPLICATION_KEY
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import com.tritiumgaming.phasmophobiaevidencepicker.app.PETApplication
import com.tritiumgaming.phasmophobiaevidencepicker.data.repository.MissionRepository
import com.tritiumgaming.phasmophobiaevidencepicker.domain.model.missions.MissionsListModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

class ObjectivesViewModel(
    missionRepository: MissionRepository
): ViewModel() {

    /* All possible objectives */
    var missionsListModel: MissionsListModel? = MissionsListModel(missionRepository.objectivesList)
    /* Objective Completed Buttons */
    var spinnerCompletionStatus: BooleanArray = BooleanArray(3) { NOT_COMPLETE }

    /* Ghost name */
    private val _ghostName = MutableStateFlow("")
    val ghostName = _ghostName.asStateFlow()
    fun setGhostName(ghostName: String? = null) {
        _ghostName.update { ghostName ?: "" }
    }

    /* Response */
    private var _responseState = MutableStateFlow<Response>(UNKNOWN) // alone , group
    val responseState = _responseState.asStateFlow()
    fun setResponseState(response: Response = UNKNOWN) {
        _responseState.update { response }
    }

    fun toggleCompletionStatus(spinnerIndex: Int) {
        spinnerCompletionStatus[spinnerIndex] = !spinnerCompletionStatus[spinnerIndex]
    }

    fun reset() {
        setGhostName()
        missionsListModel?.reset()
        spinnerCompletionStatus.all { NOT_COMPLETE }
        _responseState.update { UNKNOWN }
    }

    class ObjectivesFactory(
        private val missionRepository: MissionRepository
    ) : ViewModelProvider.Factory {

        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            if (modelClass.isAssignableFrom(ObjectivesViewModel::class.java)) {
                @Suppress("UNCHECKED_CAST")
                return ObjectivesViewModel(
                    missionRepository
                ) as T
            }
            throw IllegalArgumentException("Unknown ViewModel class")
        }
    }

    companion object {

        const val UNKNOWN: Response = 0
        const val ALONE: Response = 1
        const val GROUP: Response = 2

        const val NOT_COMPLETE: MissionStatus = false
        const val COMPLETE: MissionStatus = true


        val Factory: ViewModelProvider.Factory = viewModelFactory {
            initializer {
                val appKeyContainer = (this[APPLICATION_KEY] as PETApplication).container

                val missionRepository: MissionRepository = appKeyContainer.missionRepository

                ObjectivesViewModel(
                    missionRepository = missionRepository
                )
            }
        }
    }
}

typealias Response = Int
typealias MissionStatus = Boolean
